src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,389 @@
+  assert(pos!=std::begin(instrs));
+  const auto prev=std::prev(pos);
+
+  std::map<goto_programt::const_targett, goto_programt::targett> target_map;
+  for(auto it=instrs.begin(), instr_end=instrs.end();
+      it!=instr_end;
+      ++it)
+  {
+    const auto p=target_map[it]=
+      target.insert(pos, goto_programt::instructiont());
+    *p=*it;
+  }
+
+  for(auto i=std::next(prev); i!=pos; ++i)
+  {
+    for(auto &t : i->targets)
+    {
+      const auto new_target=target_map.find(t);
+      assert(new_target!=std::end(target_map));
+      t=new_target->second;
+    }
+  }
+
+  return pos;
+}
+
+/*******************************************************************\
+
+Function: is_symbol_with_id
+
+  Inputs:
+    expr: The expression which may be a symbol.
+    identifier: Some identifier.
+
+ Outputs: True if the expression is a symbol with the specified identifier.
+
+ Purpose: Return whether the expression is a symbol with the specified
+          identifier.
+
+\*******************************************************************/
+
+static bool is_symbol_with_id(const exprt& expr, const dstringt& identifier)
+{
+  return expr.id()==ID_symbol && expr.get(ID_identifier)==identifier;
+}
+
+/*******************************************************************\
+
+Function: is_typecast_with_id
+
+  Inputs:
+    expr: The expression which may be a typecast.
+    identifier: Some identifier.
+
+ Outputs: True if the expression is a typecast with one operand, and the
+          typecast's identifier matches the specified identifier.
+
+ Purpose: Return whether the expression is a typecast with the specified
+          identifier.
+
+\*******************************************************************/
+
+static bool is_typecast_with_id(const exprt& expr, const dstringt& identifier)
+{
+  if(!(expr.id()==ID_typecast && expr.operands().size()==1))
+  {
+    return false;
+  }
+  const auto &typecast=to_typecast_expr(expr);
+  // Return whether the typecast has the expected operand
+  return typecast.op().get(ID_identifier)==identifier;
+}
+
+/*******************************************************************\
+
+Function: is_assignment_from
+
+  Inputs:
+    instr: A goto program instruction.
+    identifier: Some identifier.
+
+ Outputs: True if the expression is a typecast with one operand, and the
+          typecast's identifier matches the specified identifier.
+
+ Purpose: Return whether the instruction is an assignment, and the rhs is a
+          symbol or typecast expression with the specified identifier.
+
+\*******************************************************************/
+
+static bool is_assignment_from(
+  const goto_programt::instructiont &instr,
+  const dstringt &identifier)
+{
+  // If not an assignment, return false
+  if(!instr.is_assign())
+  {
+    return false;
+  }
+  const auto &rhs=to_code_assign(instr.code).rhs();
+  return is_symbol_with_id(rhs, identifier) ||
+         is_typecast_with_id(rhs, identifier);
+}
```

Comment:
As this is the only place these two functions are used, why not put the code here directly?